torchrl/objectives/a2c.py

```diff
@@ -0,0 +1,148 @@
+        loss_function_type (str): loss function for the value discrepancy. Can be one of "l1", "l2" or "smooth_l1".
+        advantage_module (nn.Module): TensorDictModule used to compute tha advantage function.
+    """
+
+    def __init__(
+        self,
+        actor: ProbabilisticTensorDictModule,
+        critic: TensorDictModule,
+        advantage_key: str = "advantage",
+        advantage_diff_key: str = "value_error",
+        entropy_bonus: bool = True,
+        samples_mc_entropy: int = 1,
+        entropy_coef: float = 0.01,
+        critic_coef: float = 1.0,
+        gamma: float = 0.99,
+        loss_critic_type: str = "smooth_l1",
+        advantage_module: Optional[Callable[[TensorDictBase], TensorDictBase]] = None,
+    ):
+        super().__init__()
+        self.convert_to_functional(actor, "actor")
+        self.convert_to_functional(critic, "critic", compare_against=self.actor_params)
+        self.advantage_key = advantage_key
+        self.advantage_diff_key = advantage_diff_key
+        self.samples_mc_entropy = samples_mc_entropy
+        self.entropy_bonus = entropy_bonus and entropy_coef
+        self.register_buffer(
+            "entropy_coef", torch.tensor(entropy_coef, device=self.device)
+        )
+        self.register_buffer(
+            "critic_coef", torch.tensor(critic_coef, device=self.device)
+        )
+        self.register_buffer("gamma", torch.tensor(gamma, device=self.device))
+        self.loss_critic_type = loss_critic_type
+        self.advantage_module = advantage_module
+        if advantage_module:
+            self.advantage_module = advantage_module.to(self.device)
+
+    def reset(self) -> None:
+        pass
+
+    def get_entropy_bonus(self, dist: d.Distribution) -> torch.Tensor:
+        try:
+            entropy = dist.entropy()
+        except NotImplementedError:
+            x = dist.rsample((self.samples_mc_entropy,))
+            entropy = -dist.log_prob(x)
+        return entropy.unsqueeze(-1)
+
+    def _log_probs(
+        self, tensordict: TensorDictBase
+    ) -> Tuple[torch.Tensor, d.Distribution]:
+        # current log_prob of actions
+        action = tensordict.get("action")
+        if action.requires_grad:
+            raise RuntimeError("tensordict stored action requires grad.")
+        tensordict_clone = tensordict.select(*self.actor.in_keys).clone()
+
+        dist, *_ = self.actor.get_dist(
+            tensordict_clone, params=self.actor_params, buffers=self.actor_buffers
+        )
+        log_prob = dist.log_prob(action)
+        log_prob = log_prob.unsqueeze(-1)
+        return log_prob, dist
+
+    def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
+        if self.advantage_diff_key in tensordict.keys():
+            advantage_diff = tensordict.get(self.advantage_diff_key)
+            if not advantage_diff.requires_grad:
+                raise RuntimeError(
+                    "value_target retrieved from tensordict does not requires grad."
```

Comment:
I guess this is a copy paste so it's a self blame comment
"Requires" should read "require"